### ISSUE TITLE
Feature suggestion: format code narrowing

### ISSUE THREAD

**Issue by jonathandung** [astral-sh/ty#4046]
From astral-sh/ruff#25744.

```py
# goals in increasing order of ambition
from typing import Literal
class foo:
    # Invalid override of method `__format__`: Definition is incompatible with `object.__format__`
    def __format__(self, format_spec: Literal['a', '42'], /) -> str: ... # silence this
bar: foo
f'{bar:b}' # flag this
format(bar, 'b') # flag this
'(%b, %d)' % (bar, 1) # flag this
```

[Playground](https://play.ty.dev/d404937e-8ea9-4072-94ee-8eee7984d154)

**Comment by sharkdp** [comment]
I can understand the desire, but given that this definition of `__format__` is a clear Liskov violation (according to the annotation of `object.__format__` in typeshed), it's unclear to me if that is a viable proposal for adding stricter type checking to format specifiers.

It feels to me like something that should be discussed outside of ty's issue tracker at first. Maybe there are better solutions to this.

Closing this for now, but feel free to comment here.

**Comment by jonathandung** [comment]
python/typing#2321
